### ISSUE TITLE
How to reset the same grid but with random agent positon in Four Rooms

### ISSUE THREAD

**Issue by nikhilrayaprolu** [maximecb/gym-minigrid#150]
I realized that on reset, the grid layout changes along with the agent position. How can I fix the grid layout but change the agent position to a random start location on reset? If I set the seed, even the agent's start state gets fixed.

**Comment by maximecb** [comment]
You should be able to call `env.reset()` and then do `env.agent_pos = np.array((i, j))`

**Comment by nikhilrayaprolu** [comment]
How can I get the observation after setting the agent position?
env.step requires action, whereas env.reset cannot be used.

**Comment by maximecb** [comment]
There is a method `env.gen_obs()`:
https://github.com/maximecb/gym-minigrid/blob/master/gym_minigrid/minigrid.py#L1197

**Comment by Abc11c** [comment]
Hi @maximecb ,  

I'm also facing the same issue, I want to retain the maze grid, while only change the agents start position

When I record the video, the agent position doesn't seem to change
```
for i in range(number_of_tests):
  env.seed(i)
  obs = env.reset()
  for _ range(2): # have the same grid change agent pos only
    env.agent_pos =  env.place_agent()
    print(f"agent pos {env.agent_pos}")
    obs = env.gen_obs()
    env = RecordVideo(env, './goto/')
```

Thanks!

**Comment by pseudo-rnd-thoughts** [comment]
@Abc11c What version of minigrid are you using? 
Could you provide a minimal testing script to work with

**Comment by Abc11c** [comment]
Hi @pseudo-rnd-thoughts 

I think I was able to resolve the issue, I was in the [BABYAI fork](https://github.com/mila-iqia/babyai) earlier, now after switching to `minigrid 2.3.1`  and adapting something from the [test_script](https://github.com/Farama-Foundation/Minigrid/blob/master/tests/test_baby_ai_bot.py), did the trick.

Not sure why the same soln work on BABAYAI env though

```
from __future__ import annotations

import gymnasium as gym
import argparse

import pytest
print(gym.__file__)
from minigrid.utils.baby_ai_bot import BabyAIBot
import logging
logger = logging.getLogger(__name__)
import sys
import subprocess
import os
import time
import numpy as np
import blosc
import torch
import pdb
import random
import copy
import matplotlib.pyplot as plt
from minigrid.wrappers import FullyObsWrapper, RGBImgObsWrapper

from gymnasium.wrappers import RecordVideo

import babyai.utils as utils
parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument("--env", required=True,
                    help="name of the environment to be run (REQUIRED)")
parser.add_argument("--model", default='BOT',
                    help="name of the trained model (REQUIRED)")
parser.add_argument("--demos", default=None,
                    help="path to save demonstrations (based on --model and --origin by default)")
parser.add_argument("--episodes", type=int, default=1000000,
                    help="number of episodes to generate demonstrations for")
parser.add_argument("--valid-episodes", type=int, default=512,
                    help="number of validation episodes to generate demonstrations for")
parser.add_argument("--seed", type=int, default=0,
                    help="start random seed")
parser.add_argument("--argmax", action="store_true", default=False,
                    help="action with highest probability is selected")
parser.add_argument("--log-interval", type=int, default=100,
                    help="interval between progress reports")
parser.add_argument("--save-interval", type=int, default=2500,
                    help="interval between demonstrations saving")
parser.add_argument("--filter-steps", type=int, default=0,
                    help="filter out demos with number of steps more than filter-steps")
parser.add_argument("--on-exception", type=str, default='warn', choices=('warn', 'crash'),
                    help="How to handle exceptions during demo generation")

parser.add_argument("--job-script", type=str, default=None,
                    help="The script that launches make_agent_demos.py at a cluster.")
parser.add_argument("--jobs", type=int, default=0,
                    help="Split generation in that many jobs")

args = parser.parse_args()

# see discussion starting here: https://github.com/Farama-Foundation/Minigrid/pull/381#issuecomment-1646800992
broken_bonus_envs = {
    "BabyAI-PutNextS5N2Carrying-v0",
    "BabyAI-PutNextS6N3Carrying-v0",
    "BabyAI-PutNextS7N4Carrying-v0",
    "BabyAI-KeyInBox-v0",
}

# get all babyai envs (except the broken ones)
babyai_envs = []
for k_i in gym.envs.registry.keys():
    if k_i.split("-")[0] == "BabyAI":
        if k_i not in broken_bonus_envs:
            babyai_envs.append(k_i)
babyai_envs = ['BabyAI-KeyCorridorS6R3-v0']
# breakpoint()
def print_demo_lengths(demos):
    num_frames_per_episode = [len(demo[2]) for demo in demos]
    logger.info('Demo length: {:.3f}+-{:.3f}'.format(
        np.mean(num_frames_per_episode), np.std(num_frames_per_episode)))

@pytest.mark.parametrize("env_id", babyai_envs)
def test_bot(env_id):
    """
    The BabyAI Bot should be able to solve all BabyAI environments,
    allowing us therefore to generate demonstrations.
    """
    # Use the parameter env_id to make the environment
    print(env_id)
    env = gym.make(env_id)
    env = gym.make(env_id, render_mode="rgb_array") # for visual debugging
    

    # reset env
    curr_seed = 0

    num_steps = 240
    terminated = False
    n_episodes = 1
    count = 0
    just_crashed = False
    demos_path = utils.get_demos_path(args.demos, args.env, 'agent', False)
    demos = []
    while True:
        if count == n_episodes:
            break
        print(f"Episode {count}")
        
        terminated = False
        # env.reset(seed=curr_seed + count)
        if just_crashed:
            logger.info("reset the environment to find a mission that the bot can solve")
            env.reset(seed=curr_seed + count)
        for pos in range(2):
            terminated = False
            obs = env.reset(seed=curr_seed + count)
            env.agent_pos = env.place_agent(i = random.randint(0, env.num_cols-1), j = random.randint(0, env.num_rows-1))
            env = RecordVideo(env, f'./video_goto/video_goto_pos_{pos}')
            # breakpoint()
            print("agent pos:",env.agent_pos)
            while not terminated:
                # create expert bot
                
                expert = BabyAIBot(env)

                last_action = None
                goal =  expert.get_goal_state()
                actions = []
                # breakpoint()
                obs = obs[0]
                mission = obs["mission"]
                full_obs = env.gen_full_obs()
                # breakpoint()
                # plt.imshow(full_obs)
                goal_state = None
                cur_states = []
                next_states = []
                images = []
                action_temp = [None]*6
                for _step in range(num_steps):
                    action = expert.replan(last_action)
                    # breakpoint()
                    
                    cur_state = list(env.agent_pos) + [obs['direction']] + goal
                    cur_states.append(cur_state)
                    # action = agent.act(obs)['action']
                    
                    obs, reward, terminated, truncated, info = env.step(action)
                    last_action = action
                    images.append(full_obs)
                    # breakpoint()
                    next_state = list(env.agent_pos) + [obs['direction']] + goal
                    next_states.append(next_state)
                    actions.append(action)
                    full_obs = env.gen_full_obs()
                    obs = obs
                    # env.render()
                    # breakpoint()
                    if reward > 0:
                        demos.append((mission, blosc.pack_array(np.array(images)), cur_states, actions, next_states))
                        just_crashed = False
                    if terminated:
                        break

                # try again with a different seed
                # curr_seed += 1
        
            env.close()
        count += 1
        
    logger.info("Saving demos...")   
    # breakpoint()
    utils.save_demos(demos, demos_path)
    logger.info("{} demos saved".format(count))
    print_demo_lengths(demos[-100:])
        
if __name__ == "__main__":
    test_bot('BabyAI-GoToObjMazeS4-v0')

```
